dds/idl/field_info.cpp

```diff
@@ -8,18 +8,41 @@
 #include "field_info.h"
 
 #include <sstream>
+#include <cstring>
 
 using namespace AstTypeClassification;
 
-FieldInfo::EleLen::EleLen(const FieldInfo& af)
-  : ele_(af.as_base_)
-  , len_(af.n_elems_)
+FieldInfo::EleLen::EleLen(AST_Type* type)
+  : cls_(classify(type))
+  , len_(container_element_limit(type))
 {
+  AST_Type* const base = resolveActualType(container_base_type(type));
+  const Classification base_cls = classify(base);
+  base_cls_ = base_cls;
+  base_name_ = base->full_name();
+  if (base_cls & (CL_ARRAY | CL_SEQUENCE)) {
+    base_container_ = Container(new EleLen(base));
+  }
 }
 
 bool FieldInfo::EleLen::operator<(const EleLen& o) const
 {
-  return ele_ < o.ele_ || (ele_ == o.ele_ && len_ < o.len_);
+  if (cls_ != o.cls_) {
+    return cls_ < o.cls_;
+  }
+  if (len_ != o.len_) {
+    return len_ < o.len_;
+  }
+  if (bool(base_container_.get()) != bool(o.base_container_.get())) {
```

Comment:
I'm not sure if ACE's smart pointer template is the same way, but with the standard library one couldn't we leave out `.get()` and just have the explicit conversion to bool?